### ISSUE TITLE
nsd should not chown its PID file

### ISSUE THREAD

**Issue by orlitzky** [NLnetLabs/nsd#317]
This is a confusing one because it relies on a bunch of unwritten assumptions about how init systems work. So, apologies in advance.

There are two types of init systems these days:

* Smart ones, like OpenRC and systemd that can launch and track a process themselves, and
* Dumb ones -- pretty much only SysVinit -- that cannot.

A priori, dumb init systems have no way to track a process between when you start it and when you stop it. One unwritten assumption is that _PID files exist solely to provide dumb init systems with a way to stop the same process they started_. Another unwritten assumption is that all init systems are running as root. When the administrator runs `/etc/init.d/nsd stop`, that's root running a shell script.

Currently, nsd creates its PID file before dropping privileges (when `HAVE_SETUSERCONTEXT` is false, anyway), but then calls `chown` on the result: https://github.com/NLnetLabs/nsd/blob/master/nsd.c#L727C6-L727C11

This leads to the PID file being owned by `nsd:nsd`. This creates an unavoidable (but minor) security vulnerability. When I run `/etc/init.d/nsd stop` as root, the `nsd` user can (say) write `1` into the PID file, causing the machine to reboot. Of course there are other root processes he could kill, but that should be enough to convince you that something bad has happened.

Usually a workaround for this is to run the process in the foreground, and let OpenRC/systemd push nsd into the background after writing its own (`root:root`) PID file. However, nsd does not support running in the foreground. NSD also forks unconditionally to drop privileges, so if we try to let OpenRC/systemd record the PID themselves, we wind up with the wrong PID!

In short, no matter what we do, we are in a predicament when it comes to stopping the daemon.

I know _why_ you are `chown`ing your PID file: it's so you can clean it up before exiting. But I would encourage you to shrug and leave that up to the init systems. Smart ones do it automatically, and with SysVinit it's an `rm`.

A few similar reports (ctrl-f pid on my website for more): [CVE-2017-14102](https://michael.orlitzky.com/cves/cve-2017-14102.xhtml), [CVE-2017-11746](https://michael.orlitzky.com/cves/cve-2017-11746.xhtml), [CVE-2017-12847](https://michael.orlitzky.com/cves/cve-2017-12847.xhtml)

Usually this is a distro issue, but in this case, there's no way for a distro to work around it.

**Comment by wcawijngaards** [comment]
Yes, NSD does support running in the foreground. The `nsd -d` option makes it run in the foreground. The init script can then keep track of the process.

It is also possible to write the logs to stderr with like `logfile: "/dev/stderr"`. This is useful if the process should really run in a console. Behind that NSD forks the processes that it makes.

Perhaps this solves the issue? Since the init manager did not have NSD fork away from it with the '-d' flag. It can also be set with an option in nsd.conf, it is named `debug-mode: yes`, but that stops the fork away, and does not enable debug facilities so much.

**Comment by orlitzky** [comment]
That's embarassing, I don't know how I missed the `-d` flag! Yes, it does allow us to work around the issue, thank you.

There are still some small improvements available if you're interested, though:

1. The PID file will still be created insecurely in case someone decides to use it with SysVInit (not me, but someone).
2. NSD is still trying to create a PID file with `-d`. There's no reason for it to do this if you believe that the PID file only exists to aid an init system.

One consequence of the second item is that we're going to have two PID files, even with OpenRC/systemd. One root:root PID file managed by the init system, and one nsd:nsd file created by nsd itself.

**Comment by wcawijngaards** [comment]
It is possible to stop NSD from creating the pidfile. With the option `pidfile: ""` then no pidfile is created by NSD. Also `nsd -d -P ""` gives the option on the commandline to stop creation of a pidfile.

It is also possible to set the default pidfile at compile time with `./configure --with-pidfile=""` and then the pidfile is not created, unless the commandline flag or option is used to set a pidfile.

**Comment by orlitzky** [comment]
Even better, thanks :)

**Comment by wcawijngaards** [comment]
It is nice that it solves the problem! This issue can be closed.

**Comment by orlitzky** [comment]
Just to be clear, I'm only able to fix the vulnerability on Gentoo. Our existing OpenRC and systemd service scripts both let the daemon fork and rely on the PID file it creates to stop the process later on:

*  https://gitweb.gentoo.org/repo/gentoo.git/tree/net-dns/nsd/files/nsd.initd-r1
* https://gitweb.gentoo.org/repo/gentoo.git/tree/net-dns/nsd/files/nsd.service

This is exploitable by the "nsd" user to e.g. reboot the machine. but can be fixed by updating those scripts to run NSD in the foreground without a PID file, and by letting OpenRC/systemd manage it. (This is an easy solution only because I am a Gentoo developer.)

Anyone else relying on the PID file to stop the daemon outside of Gentoo is going to have the same issue though. It looks like the systemd script you have under `contrib/` is safe, but the init script you provide is vulnerable to the same issue since it calls `kill` on the contents of the PID file: https://github.com/NLnetLabs/nsd/blob/master/contrib/nsd.init

**Comment by wcawijngaards** [comment]
So you NSD wants to remove the pidfile, and it chowns to have that access. I understand you can solve the issue for yourself. But now you want this also to be fixed for, other, init system users.

**Comment by wcawijngaards** [comment]
So a problem is that the pidfile should be removed after the server stops and the chown is needed to do that. And you want it to not do that, to restrict the access of NSD. NSD supports chroot, and I think you should enable that if you want the server process to have restricted access. If the pidfile was located outside of the chroot NSD could not access it, despite this chown call.

So what would be solution look like for the init script? I want to remove the pidfile afterwards and then detect if the server was closed correctly on the next start. I guess it would be solved if the init system scripted its access to the pidfile differently? Or checked the pid it got from the pidfile before using it? That it runs a process owned by nsd? Eg. sudo -u 'nsd' kill the_pid or something along those lines?

Another way is to create an option to select for when the user wants things set up in this manner, some sort of 'disable-chown-for-pidfile' option. But that needs the user to set it, and I then do not know if it would actually be considered useful by init system users? You mean that for this particular security reason the chown should be remove plain up and NSD should be littering leftover pidfiles?

**Comment by orlitzky** [comment]
I'm not going to hound you over it, but yes, it might still be a problem for others. "Dumb" init systems like SysVinit need daemons to fork, drop privileges, and write a PID file for a few reasons:

* There's no (POSIX) standard way to launch a process with reduced privileges.
* When stopping the daemon, you want to be very sure that you stop only the process that you started. For example, you don't want to stop all "nsd" processes on the machine with `/etc/init.d/nsd stop`. You only want to stop the one (if any) that was started with `/etc/init.d/nsd start` earlier.
* Forking changes the PID, so the init system itself can't always figure out what the correct PID is.

It's also important to note that all of these init systems run as root. So if you are using SysVinit and run `/etc/init.d/nsd stop`, it is sending a signal to the process in the PID file as root. The problem is, then: after the `chown`, the nsd user controls what process is killed by root. He can abuse this to kill the firewall, for example, or reboot the machine.

Modern init systems can work around the problem by running in the foreground, but anyone using an older init system (the _only_ reason to have a PID file in the first place) is unable to safely stop NSD. This could be fixed by omitting the `chown`, at the expense of either leaving the PID around, or having the init script clean up after a shutdown.

**Comment by wcawijngaards** [comment]
I agree it is a good idea to safely deal with the pidfile privileges. The init script performing a clean up is a good idea.

**Comment by orlitzky** [comment]
> So a problem is that the pidfile should be removed after the server stops and the chown is needed to do that. And you want it to not do that, to restrict the access of NSD. NSD supports chroot, and I think you should enable that if you want the server process to have restricted access. If the pidfile was located outside of the chroot NSD could not access it, despite this chown call.

On most operating systems, `chroot()` isn't considered a security feature (e.g. [linux](https://lkml.org/lkml/2007/9/25/434)). I'm not an expert on this, so I have to defer to google. In any case, if NSD is able to delete the PID file, it can write to it. So by making it secure in this manner we would _also_ be unable to delete the PID file after shutting down.

> So what would be solution look like for the init script? I want to remove the pidfile afterwards and then detect if the server was closed correctly on the next start. I guess it would be solved if the init system scripted its access to the pidfile differently

The init script can remove the PID file after it stops the daemon successfully.

> Or checked the pid it got from the pidfile before using it? That it runs a process owned by nsd? Eg. sudo -u 'nsd' kill the_pid or something along those lines?

In the past I've looked for a solution like this, but I never found one that is reliable. You can't count on `sudo` being installed, and `pkill` has the same problem.  There's no POSIX command to drop privileges and run a command. Even the critical arguments to  `su` vary across operating systems.

> Another way is to create an option to select for when the user wants things set up in this manner, some sort of 'disable-chown-for-pidfile' option. But that needs the user to set it, and I then do not know if it would actually be considered useful by init system users? You mean that for this particular security reason the chown should be remove plain up and NSD should be littering leftover pidfiles?

Essentially, yes. When OpenRC or systemd are used with the forking daemon and a PID file, they will clean it up automatically. The only real problem is with old SysVinit. However, there it is fair game to require the init script itself to remove the PID file after a clean shutdown.

If the PID files that are created with `nsd -d` are annoying, they are also unnecessary and could simply be disabled by `-d`.

**Comment by wcawijngaards** [comment]
The messages seem to cross each other several times now. Annoying that privdrop for the kill would not work. But remove of pidfile by init script sounds fine with me as well.

I would not really want to change the existing behaviour of the `-d` flag, since there is already another flag for it.

**Comment by orlitzky** [comment]
Yes, and honestly, leaving an unused, 5-byte file laying around in a temporary directory is not the end of the world.

**Comment by wcawijngaards** [comment]
The fix commit, adds a line in the contrib init script that removes the pidfile after it was shut down by the init script. Also it removes the chown call, of course. And NSD no longer checks for the old pid at startup, and does not print an error when it cannot remove the pidfile at exit. The printout is visible at high verbosity levels. The check is only disabled when there is a username configured, without one NSD assumes the pidfile could not be there, and then could print a warning.

I guess that should fix the issue reported. Thanks for the report and detailed explanation!

**Comment by orlitzky** [comment]
Yes, and thank you for the quick fix.

**Comment by orlitzky** [comment]
As a follow-up, you may want to add `-P ""` to the systemd service script now: https://github.com/NLnetLabs/nsd/blob/master/contrib/nsd.service#L7

Otherwise a tmpfiles.d entry is needed to create `/run/nsd` to hold the PID file, but systemd won't be using that PID file anyway.

My migration to nsd went smoothly by the way :)

**Comment by wcawijngaards** [comment]
The commit https://github.com/NLnetLabs/nsd/commit/1366b6a9f3528ab6438013be8a62ddd3e64a210f adds the commandline parameter. Nice that things are working well!

**Comment by lachesis** [comment]
Lame, this broke my system. I was using this pid file to HUP nsd when I made changes to its config, but the new systemd settings override the pid_file entry in my config and made the pid file not get created.

**Comment by orlitzky** [comment]
> Lame, this broke my system. I was using this pid file to HUP nsd when I made changes to its config, but the new systemd settings override the pid_file entry in my config and made the pid file not get created.

The systemd service file already knows how to reload the daemon: https://github.com/NLnetLabs/nsd/blob/master/contrib/nsd.service#L8

Instead of HUPing the daemon manually, try `systemctl reload`.

**Comment by orlitzky** [comment]
...and if for some reason `systemctl reload` won't work, then `$NSD_EXTRA_OPTS` are still appended to the end of the command, allowing you to pass your own `-P` flag in: https://github.com/NLnetLabs/nsd/blob/master/contrib/nsd.service#L7

**Comment by lachesis** [comment]
Thanks for the response. `systemctl reload` will not work for three reasons:

1. My project doesn't depend on systemd.
2. My project doesn't run any shell commands. Instead it uses `process.kill` in NodeJS.
3. My project runs as the `nsd` user not the `root` user, so it can't call `systemctl reload`.

So the recommended workaround is to edit `/etc/default/nsd` and add `-P "/run/nsd/nsd.pid"` to the end of the NSD_EXTRA_OPTS variable?

**Comment by za3k** [comment]
(on the same project as lachesis) Thanks for the help. 

That said, I think a default service file making a documented config variable be (silently) ignored is IMO a bug. I would change the default config file instead of the command line options.

If you do need the command line default for some reason, it would be nice to document this gotcha anywhere `pidfile` is mentioned like the nsd.conf man page. Or, consider removing the config option?

**Comment by orlitzky** [comment]
> Thanks for the response. `systemctl reload` will not work for three reasons:
> 
>     1. My project doesn't depend on systemd.
> 
>     2. My project doesn't run any shell commands. Instead it uses `process.kill` in NodeJS.
> 
>     3. My project runs as the `nsd` user not the `root` user, so it can't call `systemctl reload`.

Ok, fair enough.

> So the recommended workaround is to edit `/etc/default/nsd` and add `-P "/run/nsd/nsd.pid"` to the end of the NSD_EXTRA_OPTS variable?

That would be my suggestion for now. I recommended the `-P ""` in https://github.com/NLnetLabs/nsd/issues/317#issuecomment-2041438207 because systemd isn't using the PID file and creating one can cause headaches. Before nsd can write its own PID file, the directory it will live in has to be created. Typically that will be something like `/run/nsd`, and the way that you would create that is... through systemd, with a [tmpfiles](https://www.freedesktop.org/software/systemd/man/latest/tmpfiles.d.html) entry. But the tmpfiles entry is static and has no idea what's inside your `nsd.conf`. Your linux distro can set matching defaults in both `nsd.conf` and the tmpfiles entry out of the box, but as soon as you edit the PID file path in `nsd.conf`, it all falls apart. You can wind up with a daemon that won't start because it can't create a PID file that won't even be used!

Anyway, that's my digression. The reason I would work around this with `NSD_EXTRA_OPTS` is because, ultimately, you can't control how nsd will be started. In this case it sounds like you're using a distro where the upstream service file is used -- that both causes your problem and provides you with `NSD_EXTRA_OPTS` as a solution. Other distros might have custom service files that cause new and exciting problems and the general solution in those situations would be to "fix the systemd unit file to not be dumbshit." :) In other words, users on those systems will have to mess with the systemd config, or the config of whatever init system is being used.

But with systemd and whenever the upstream nsd service file is used, `NSD_EXTRA_OPTS` should work. Plus, setting it under any other circumstances is harmless. So it solves only this problem, on distros that have the problem, and is using the same general mechanism that would be used to solve other problems of this type.

**Comment by orlitzky** [comment]
> That said, I think a default service file making a documented config variable be (silently) ignored is IMO a bug. I would change the default config file instead of the command line options.

I mostly agree but the underlying problem is that specifying the PID file location in a config file can never work properly (my previous comment). You can put `/run/foo/nsd.pid` in there, but it won't work. To make it work, you end up messing with the init system; but if you're going to mess with the init system, then you might as well just make it do what you want in the first place.

> Or, consider removing the config option?

Ultimately I think this is the right thing to do because the PID file is necessarily the domain of the init system. The trick is to get there incrementally without breaking anyone's setup or annoying users to death with constant changes.

**Comment by wcawijngaards** [comment]
I added a commit that explains more about the pidfile permissions in the man page. Not sure where documentation on the init system would go, it depends on whether the contrib/nsd.init or contrib/nsd.service sample file, or perhaps a different init system setup is used. The man page text could go into detail on each of those cases with advice, if needed.

Removal of the pidfile config option does not seem to be a good idea to me. That would annoy users that have it. For example, on a system where nsd is started from rc.local, or the nsd server is set up in user homedir. Also there is a set up with like  `/etc/nsd.d/pidfile.conf`, with an `include: "/etc/nsd.d/*.conf"` in the nsd.conf file. Then the file with the init system directive and files with user options can coexist.

That said, there could be documentation on the option when it is not in use, or a configure option that disables config file settings.
